ow_dynamic_terrain/src/DynamicTerrainVisual.cpp

```diff
@@ -78,6 +88,53 @@ class DynamicTerrainVisual : public VisualPlugin, public DynamicTerrainBase
   {
     onModifyTerrainMsg(msg, TerrainModifier::modifyPatch);
   }
+
+  void applyNormalMap()
+  {
+    auto heightmap = getHeightmap(get_scene());
+    if (heightmap == nullptr)
+    {
+      gzerr << m_plugin_name << ": DynamicTerrainVisual::applyNormalMap()"
+        " could not acquire heightmap." << endl;
+      return;
+    }
+    auto terrain = heightmap->OgreTerrain()->getTerrain(0, 0);
+
+    // Override custom material normal map labeled "ow_dynamic_terrain_normal_map"
+    // with Ogre::Terrain's internal normal map.
+    // Ogre::Terrain generates an internal normal map and it is updated by
+    // Terrain::updateDerivedData(). But if you use a custom material script
+    // with a custom normal map, the internal normal map will never be seen.
+    // Note: If you are not using a custom material this code will have no
+    // effect, but Ogre should be using the correct normal map and a generated
+    // material and shader in that case.
+    // Caveat: The internal normal map is always sized to the terrain's DEM
+    // resolution, so using a larger texture for more detail is not possible
+    // with this plugin.
+    ResourceManager::ResourceMapIterator res_it = MaterialManager::getSingleton().getResourceIterator();
+    while (res_it.hasMoreElements())
+    {
+      ResourcePtr resource = res_it.getNext();
+      MaterialPtr material = resource.staticCast<Material>();
+      Material::TechniqueIterator tech_it = material->getTechniqueIterator();
+      while (tech_it.hasMoreElements())
+      {
+        Technique* technique = tech_it.getNext();
+        Technique::PassIterator pass_it = technique->getPassIterator();
+        while (pass_it.hasMoreElements())
+        {
+          Pass* pass = pass_it.getNext();
+          TextureUnitState* tus = pass->getTextureUnitState("ow_dynamic_terrain_normal_map");
```

Comment:
[optional] These declarations could be `auto`, but I'm not picky about it :shrug:

Reply:
I prefer to only use auto to replace really complex or unknowable types. Showing types usually makes code easier to understand, in my opinion.